### PR TITLE
[ao][sparsity] Implemented state_dict() and load_state_dict() functions

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79883
* #79822
* #79817

The state_dict() of the DataScheduler contains all the class attributes of
the scheduler other than the data_sparsifier.
Remember to store and restore the state_dict() of the data_sparsifier along
with the scheduler

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataScheduler```

Differential Revision: [D37358607](https://our.internmc.facebook.com/intern/diff/D37358607)

### CODE REVIEW

**Review by HDCharles — APPROVED** [review]
lgtm

**Review by z-a-f — APPROVED** [review]
